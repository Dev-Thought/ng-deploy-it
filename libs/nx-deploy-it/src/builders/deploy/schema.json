{
  "$schema": "https://json-schema.org/draft-07/schema",
  "$id": "https://json-schema.org/draft-07/schema",
  "title": "NxDeployIt Deploy infrastructure",
  "description": "Deploy the infrastructure for the app",
  "type": "object",
  "properties": {
    "noBuild": {
      "type": "boolean",
      "default": false,
      "description": "Skip build process during deployment"
    },
    "nonInteractive": {
      "type": "boolean",
      "default": false,
      "description": "Changes will apply without user input"
    }
  },
  "required": []
}
