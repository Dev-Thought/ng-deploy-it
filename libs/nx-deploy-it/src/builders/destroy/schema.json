{
  "$schema": "https://json-schema.org/draft-07/schema",
  "$id": "https://json-schema.org/draft-07/schema",
  "title": "NxDeployIt Destroy infrastructure",
  "description": "Destroy the infrastructure for the app",
  "type": "object",
  "properties": {
    "nonInteractive": {
      "type": "boolean",
      "default": false,
      "description": "Changes will apply without user input"
    }
  },
  "required": []
}
