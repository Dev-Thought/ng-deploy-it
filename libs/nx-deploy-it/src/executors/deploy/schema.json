{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "title": "NxDeployIt Deploy infrastructure",
  "description": "Deploy the infrastructure for the app",
  "type": "object",
  "properties": {
    "nonInteractive": {
      "type": "boolean",
      "default": false,
      "description": "Changes will apply without user input"
    }
  },
  "required": []
}
