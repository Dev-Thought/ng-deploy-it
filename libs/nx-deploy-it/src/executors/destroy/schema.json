{
  "$schema": "http://json-schema.org/schema",
  "title": "NxDeployIt Destroy infrastructure",
  "description": "Destroy the infrastructure for the app",
  "type": "object",
  "properties": {
    "nonInteractive": {
      "type": "boolean",
      "default": false,
      "description": "Changes will apply without user input"
    }
  },
  "required": []
}
