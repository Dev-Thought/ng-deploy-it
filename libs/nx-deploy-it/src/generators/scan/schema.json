{
  "$schema": "http://json-schema.org/draft-07/schema",
  "id": "scan",
  "title": "",
  "type": "object",
  "cli": "nx",
  "properties": {}
}
