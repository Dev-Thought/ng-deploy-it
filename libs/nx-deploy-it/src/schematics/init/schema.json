{
  "$schema": "http://json-schema.org/schema",
  "id": "NxDeployItInit",
  "title": "Add nx-deploy-it cloud configuration for an application",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "provider": {
      "type": "string",
      "description": "Your cloud provider",
      "x-prompt": {
        "message": "Please choose your provider",
        "type": "list",
        "items": [
          {
            "label": "AWS",
            "value": "aws"
          },
          {
            "label": "Azure",
            "value": "azure"
          },
          {
            "label": "Google Cloud Platform",
            "value": "gcp"
          }
        ]
      }
    },
    "customDomainName": {
      "type": "string",
      "description": "Your custom domain which will be mapped to the static website / cdn"
    }
  },
  "required": ["provider"]
}
