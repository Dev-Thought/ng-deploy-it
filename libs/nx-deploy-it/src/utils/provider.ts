export enum PROVIDER {
  AWS = 'aws',
  AZURE = 'azure',
  GOOGLE_CLOUD_PLATFORM = 'gcp',
}

export function getCloudTemplateName(cloudProvider: string): string {
  return `${cloudProvider}-typescript`;
}
