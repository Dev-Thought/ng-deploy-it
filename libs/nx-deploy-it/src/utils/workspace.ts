import { resolve } from 'path';
import {
  ProjectDefinition,
  WorkspaceDefinition
} from '@angular-devkit/core/src/workspace';
import { BaseAdapter } from '../schematics/init/adapter/base.adapter';
import { getApplicationType, ApplicationType } from './application-type';
import { NxDeployItInitSchematicSchema } from '../schematics/init/schema';
import { WebappAdapter } from '../schematics/init/adapter/webapp.adapter';
import { NestJSAdapter } from '../schematics/init/adapter/nestjs.adapter';

export function getRealWorkspacePath() {
  // TODO!: find a better way
  return process.cwd();
}

export function getPulumiBinaryPath() {
  return resolve(getRealWorkspacePath(), 'node_modules/.bin/pulumi');
}

export function getAdapter(
  project: ProjectDefinition,
  options: NxDeployItInitSchematicSchema
): BaseAdapter {
  const applicationType = getApplicationType(project.targets.get('build'));
  switch (applicationType) {
    case ApplicationType.ANGULAR:
    case ApplicationType.REACT:
      return new WebappAdapter(project, options, applicationType);
    case ApplicationType.NESTJS:
      return new NestJSAdapter(project, options, applicationType);
    default:
  }

  throw new Error(
    `Can't recognize application type. Supported list can be found here: https://github.com/Dev-Thought/nx-plugins/libs/nx-deploy-it`
  );
}

export function getApplications(
  workspace: WorkspaceDefinition
): { projectName: string; applicationType: ApplicationType }[] {
  const applications: {
    projectName: string;
    applicationType: ApplicationType;
  }[] = [];
  workspace.projects.forEach((project, projectName) => {
    const applicationType = getApplicationType(project.targets.get('build'));
    if (applicationType) {
      applications.push({
        projectName,
        applicationType
      });
    }
  });

  return applications;
}
