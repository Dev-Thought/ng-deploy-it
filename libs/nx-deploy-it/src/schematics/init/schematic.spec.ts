import { Tree } from '@angular-devkit/schematics';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { createEmptyWorkspace } from '@nrwl/workspace/testing';
import { join } from 'path';
import { NxDeployItInitSchematicSchema } from './schema';
import { readJsonInTree } from '@nrwl/workspace';
import { stdin } from 'mock-stdin';

import { PROVIDER } from '../../utils/provider';
import { createPulumiMockProjectInTree } from '../../utils-test/pulumi.mock';
import * as childProcess from 'child_process';
import * as fs from 'fs';
import { createApplication } from '../../utils-test/app.utils';
import {
  answerInitQuestionsAWS,
  answerInitQuestionsAzure,
  answerInitQuestionsGCP,
} from '../../utils-test/enquirer.utils';

describe('init schematic', () => {
  let appTree: Tree;
  const projectName = 'mock-project';

  const unlinkSync = jest.spyOn(fs, 'unlinkSync');
  const spawnSync = jest.spyOn(childProcess, 'spawnSync');

  const originReadFileSync = fs.readFileSync;

  (fs.readFileSync as any) = jest
    .fn(originReadFileSync)
    .mockImplementation((path, options) => {
      if (path === `apps/${projectName}/infrastructure/Pulumi.yaml`) {
        return '';
      }

      return originReadFileSync(path, options);
    });

  const testRunner = new SchematicTestRunner(
    '@dev-thought/nx-deploy-it',
    join(__dirname, '../../../collection.json')
  );

  let io = null;
  beforeAll(() => (io = stdin()));
  afterAll(() => io.restore());

  beforeEach(() => {
    appTree = createEmptyWorkspace(Tree.empty());
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('pulumi project', () => {
    const options: NxDeployItInitSchematicSchema = {
      project: projectName,
      provider: PROVIDER.AWS,
    };

    beforeEach(async () => {
      appTree = await createApplication(
        testRunner,
        projectName,
        'nest',
        appTree
      );

      spawnSync.mockImplementation(() => {
        createPulumiMockProjectInTree(appTree, PROVIDER.AWS, projectName);
        return {} as any;
      });
      unlinkSync.mockImplementation();
    });

    it('should be initialized', async () => {
      answerInitQuestionsAWS(io, null, null);

      const tree = await testRunner
        .runSchematicAsync('init', options, appTree)
        .toPromise();

      expect(
        tree.exists(`apps/${projectName}/infrastructure/Pulumi.yaml`)
      ).toBeTruthy();

      expect(spawnSync).toHaveBeenCalled();
      expect(spawnSync.mock.calls[0][1][1]).toContain('aws-typescript');
      expect(spawnSync.mock.calls[0][1][3]).toBe(projectName);
      expect(spawnSync.mock.calls[0][1][5]).toContain(
        `apps/${projectName}/infrastructure`
      );
      expect(spawnSync.mock.calls[0][1][7]).toContain(
        'Infrastructure as Code based on Pulumi - managed by @dev-thought/nx-deploy-it'
      );
      expect(unlinkSync).toHaveBeenCalledTimes(5);
    });

    it('should fail if the project already has an deploy integration', async () => {
      answerInitQuestionsAWS(io, null, null);

      await testRunner.runSchematicAsync('init', options, appTree).toPromise();
      await testRunner.runSchematicAsync('init', options, appTree).toPromise();

      expect(spawnSync).toHaveBeenCalledTimes(1);
    });
  });

  describe('aws provider', () => {
    const options: NxDeployItInitSchematicSchema = {
      project: projectName,
      provider: PROVIDER.AWS,
    };

    beforeEach(async () => {
      appTree = await createApplication(
        testRunner,
        projectName,
        'nest',
        appTree
      );

      spawnSync.mockImplementation(() => {
        createPulumiMockProjectInTree(appTree, PROVIDER.AWS, projectName);
        return {} as any;
      });
      unlinkSync.mockImplementation();
    });

    it('should add dependencies to package.json', async () => {
      answerInitQuestionsAWS(io, null, null);

      spawnSync.mockImplementation(() => {
        createPulumiMockProjectInTree(appTree, PROVIDER.AWS, projectName);
        return {} as any;
      });
      unlinkSync.mockImplementation();
      const tree = await testRunner
        .runSchematicAsync('init', options, appTree)
        .toPromise();

      const packageJSON = readJsonInTree(tree, 'package.json');
      expect(packageJSON.dependencies['@pulumi/pulumi']).toBeDefined();
      expect(packageJSON.dependencies['@pulumi/aws']).toBeDefined();
      expect(packageJSON.dependencies['@pulumi/awsx']).toBeDefined();
    });

    it('should extend the project configuration options with aws profile', async () => {
      answerInitQuestionsAWS(io, 'eu-central-1', 'my-aws-profile');

      const tree = await testRunner
        .runSchematicAsync('init', options, appTree)
        .toPromise();

      const workspaceJson = readJsonInTree(tree, 'workspace.json');
      expect(
        workspaceJson.projects[projectName].architect.deploy
      ).toMatchSnapshot();
      expect(
        workspaceJson.projects[projectName].architect.destroy
      ).toMatchSnapshot();
    });

    it('should extend the project default configuration options', async () => {
      answerInitQuestionsAWS(io, 'eu-central-1');

      const tree = await testRunner
        .runSchematicAsync('init', options, appTree)
        .toPromise();

      const workspaceJson = readJsonInTree(tree, 'workspace.json');
      expect(
        workspaceJson.projects[projectName].architect.deploy
      ).toMatchSnapshot('Deploy Action');
      expect(
        workspaceJson.projects[projectName].architect.destroy
      ).toMatchSnapshot('Destroy Action');
    });
  });

  describe('azure provider', () => {
    const options: NxDeployItInitSchematicSchema = {
      project: projectName,
      provider: PROVIDER.AZURE,
    };

    beforeEach(async () => {
      appTree = await createApplication(
        testRunner,
        projectName,
        'nest',
        appTree
      );

      spawnSync.mockImplementation(() => {
        createPulumiMockProjectInTree(appTree, PROVIDER.AZURE, projectName);
        return {} as any;
      });
      unlinkSync.mockImplementation();
    });

    it('should add dependencies to package.json', async () => {
      answerInitQuestionsAzure(io, null);

      const tree = await testRunner
        .runSchematicAsync('init', options, appTree)
        .toPromise();

      const packageJSON = readJsonInTree(tree, 'package.json');
      expect(packageJSON.dependencies['@pulumi/pulumi']).toBeDefined();
      expect(packageJSON.dependencies['@pulumi/azure']).toBeDefined();
    });

    it('should extend the project default configuration options', async () => {
      answerInitQuestionsAzure(io, 'eastasia');

      const tree = await testRunner
        .runSchematicAsync('init', options, appTree)
        .toPromise();

      const workspaceJson = readJsonInTree(tree, 'workspace.json');
      expect(
        workspaceJson.projects[projectName].architect.deploy
      ).toMatchSnapshot('Deploy Action');
      expect(
        workspaceJson.projects[projectName].architect.destroy
      ).toMatchSnapshot('Destroy Action');
    });
  });

  describe('google cloud platform provider', () => {
    const options: NxDeployItInitSchematicSchema = {
      project: projectName,
      provider: PROVIDER.GOOGLE_CLOUD_PLATFORM,
    };

    beforeEach(async () => {
      appTree = await createApplication(
        testRunner,
        projectName,
        'nest',
        appTree
      );

      spawnSync.mockImplementation(() => {
        createPulumiMockProjectInTree(
          appTree,
          PROVIDER.GOOGLE_CLOUD_PLATFORM,
          projectName
        );
        return {} as any;
      });
      unlinkSync.mockImplementation();
    });

    it('should add dependencies to package.json', async () => {
      answerInitQuestionsGCP(io, 'my-google-project-id', 'europe-west1');

      const tree = await testRunner
        .runSchematicAsync('init', options, appTree)
        .toPromise();

      const packageJSON = readJsonInTree(tree, 'package.json');
      expect(packageJSON.dependencies['@pulumi/pulumi']).toBeDefined();
      expect(packageJSON.dependencies['@pulumi/gcp']).toBeDefined();
    });

    it('should extend the project default configuration options', async () => {
      answerInitQuestionsGCP(io, 'my-google-project-id', 'europe-west1');

      const tree = await testRunner
        .runSchematicAsync('init', options, appTree)
        .toPromise();

      const workspaceJson = readJsonInTree(tree, 'workspace.json');
      expect(
        workspaceJson.projects[projectName].architect.deploy
      ).toMatchSnapshot('Deploy Action');
      expect(
        workspaceJson.projects[projectName].architect.destroy
      ).toMatchSnapshot('Destroy Action');
    });
  });
});
