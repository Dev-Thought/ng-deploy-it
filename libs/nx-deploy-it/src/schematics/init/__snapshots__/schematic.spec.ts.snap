// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`init schematic aws provider should extend the project configuration options with aws profile 1`] = `
Object {
  "builder": "@dev-thought/nx-deploy-it:deploy",
  "options": Object {
    "main": "apps/mock-project/infrastructure/index.ts",
    "provider": "aws",
    "pulumi": Object {
      "aws:profile": "my-aws-profile",
      "aws:region": "eu-central-1",
    },
  },
}
`;

exports[`init schematic aws provider should extend the project configuration options with aws profile 2`] = `
Object {
  "builder": "@dev-thought/nx-deploy-it:destroy",
  "options": Object {
    "main": "apps/mock-project/infrastructure/index.ts",
    "provider": "aws",
  },
}
`;

exports[`init schematic aws provider should extend the project default configuration options: Deploy Action 1`] = `
Object {
  "builder": "@dev-thought/nx-deploy-it:deploy",
  "options": Object {
    "main": "apps/mock-project/infrastructure/index.ts",
    "provider": "aws",
    "pulumi": Object {
      "aws:profile": "",
      "aws:region": "eu-central-1",
    },
  },
}
`;

exports[`init schematic aws provider should extend the project default configuration options: Destroy Action 1`] = `
Object {
  "builder": "@dev-thought/nx-deploy-it:destroy",
  "options": Object {
    "main": "apps/mock-project/infrastructure/index.ts",
    "provider": "aws",
  },
}
`;

exports[`init schematic azure provider should extend the project default configuration options: Deploy Action 1`] = `
Object {
  "builder": "@dev-thought/nx-deploy-it:deploy",
  "options": Object {
    "main": "apps/mock-project/infrastructure/index.ts",
    "provider": "azure",
    "pulumi": Object {
      "azure:location": "eastasia",
    },
  },
}
`;

exports[`init schematic azure provider should extend the project default configuration options: Destroy Action 1`] = `
Object {
  "builder": "@dev-thought/nx-deploy-it:destroy",
  "options": Object {
    "main": "apps/mock-project/infrastructure/index.ts",
    "provider": "azure",
  },
}
`;

exports[`init schematic google cloud platform provider should extend the project default configuration options: Deploy Action 1`] = `
Object {
  "builder": "@dev-thought/nx-deploy-it:deploy",
  "options": Object {
    "main": "apps/mock-project/infrastructure/index.ts",
    "provider": "gcp",
    "pulumi": Object {
      "gcp:project": "my-google-project-id",
      "gcp:region": "europe-west1",
    },
  },
}
`;

exports[`init schematic google cloud platform provider should extend the project default configuration options: Destroy Action 1`] = `
Object {
  "builder": "@dev-thought/nx-deploy-it:destroy",
  "options": Object {
    "main": "apps/mock-project/infrastructure/index.ts",
    "provider": "gcp",
  },
}
`;
